File: server/src/services/tools/read_console.py
"""read_console tool — pull recent Editor.log/info/warn/error entries."""

from __future__ import annotations

from typing import Annotated, Any, Literal

from fastmcp import Context

from services.registry import cocos_mcp_tool
from services.tools._common import call_bridge


@cocos_mcp_tool(
    description=(
        "Read or clear the captured Cocos Creator editor console. The "
        "extension keeps a 500-entry ring buffer of Editor.log/info/warn/"
        "error lines. After modifying scripts/scenes, call this to look "
        "for compile errors or runtime warnings."
    ),
)
async def read_console(
    ctx: Context,
    action: Annotated[
        Literal["get", "clear"],
        "Action — 'get' returns entries, 'clear' empties the buffer.",
    ] = "get",
    levels: Annotated[
        list[Literal["log", "info", "warn", "error"]] | None,
        "Filter by log level. Defaults to all levels.",
    ] = None,
    contains: Annotated[
        str | None,
        "Only return entries whose message contains this substring.",
    ] = None,
    count: Annotated[
        int,
        "Max entries returned (clamped to 500). Default 50.",
    ] = 50,
    since: Annotated[
        int | None,
        "Sequence cursor from a previous response (returns only newer entries).",
    ] = None,
) -> dict[str, Any]:
    params: dict[str, Any] = {"action": action}
    if levels is not None:
        params["levels"] = levels
    if contains is not None:
        params["contains"] = contains
    if count is not None:
        params["count"] = int(count)
    if since is not None:
        params["since"] = int(since)
    return await call_bridge("read_console", params)

File: extension/handlers/read_console.js
'use strict';

/**
 * read_console — return recent Editor console lines from the ring buffer
 * installed by main.js, optionally filtered by level/keyword/since-cursor.
 *
 * Params:
 *   action      "get" | "clear"   (default "get")
 *   levels      array of "log" | "info" | "warn" | "error"  (default all)
 *   contains    substring filter
 *   count       max entries (default 50, max 500)
 *   since       integer cursor returned by a previous call (newer-than)
 *
 * Response (get):
 *   { entries: [{seq, timestamp, level, message}, ...], nextCursor }
 */

async function handle(params, ctx) {
    const buffer = ctx && ctx.consoleBuffer;
    if (!buffer) {
        return {
            entries: [],
            nextCursor: 0,
            note: 'console buffer not initialized — bridge may not be running',
        };
    }

    const action = (params && params.action) || 'get';
    if (action === 'clear') {
        buffer.clear();
        return { cleared: true };
    }
    if (action !== 'get') {
        throw new Error('unknown action: ' + action);
    }

    const levels = Array.isArray(params && params.levels) ? params.levels : null;
    const contains = (params && typeof params.contains === 'string') ? params.contains : null;
    const since = (params && typeof params.since === 'number') ? params.since : undefined;
    let count = (params && Number.isFinite(params.count)) ? Math.floor(params.count) : 50;
    if (count <= 0) count = 50;
    if (count > 500) count = 500;

    let entries = buffer.filter({ levels, contains, since });
    // newest at the end; trim to the last `count`
    if (entries.length > count) entries = entries.slice(entries.length - count);

    const nextCursor = entries.length ? entries[entries.length - 1].seq : (since || 0);

    return { entries, nextCursor, totalBuffered: buffer.all().length };
}

module.exports = { name: 'read_console', handle };
